### PR TITLE
Implement fork choice spec tests

### PR DESCRIPTION
**Pull request by terencechain**
This PR implements fork choice spec tests for phase 0 and Altair. Bellatrix will come later.
While implementing fork choice spec tests, we discovered some code still had old fork choice spec logic, but they were insubstantial, and this PR removed them.
We also provided a few getters from blockchain service pkg to allow fork choice spec test to retrieve and verify.

### CODE REVIEW

File: beacon-chain/blockchain/chain_info.go [hunk 1]
```diff
@@ -107,6 +107,16 @@ func (s *Service) PreviousJustifiedCheckpt() *ethpb.Checkpoint {
 	return ethpb.CopyCheckpoint(cp)
 }
 
+// BestJustifiedCheckpt returns the best justified checkpoint from store.
+func (s *Service) BestJustifiedCheckpt() *ethpb.Checkpoint {
```
Comment (terencechain) [inline beacon-chain/blockchain/chain_info.go:111]: Used for fork choice spec tests to verify against specs

File: beacon-chain/blockchain/head.go [hunk 2]
```diff
@@ -23,6 +23,24 @@ import (
 	"go.opencensus.io/trace"
 )
 
+// UpdateHead updates the beacon state head.
+func (s *Service) UpdateHead(ctx context.Context) error {
```
Comment (terencechain) [inline beacon-chain/blockchain/head.go:27]: Used by fork choice spec tests to update head

File: beacon-chain/blockchain/new_slot.go [hunk 3]
```diff
@@ -29,7 +29,7 @@ import (
 //        ancestor_at_finalized_slot = get_ancestor(store, store.best_justified_checkpoint.root, finalized_slot)
 //        if ancestor_at_finalized_slot == store.finalized_checkpoint.root:
 //            store.justified_checkpoint = store.best_justified_checkpoint
-func (s *Service) newSlot(ctx context.Context, slot types.Slot) error {
+func (s *Service) NewSlot(ctx context.Context, slot types.Slot) error {
```
Comment (terencechain) [inline beacon-chain/blockchain/new_slot.go:32]: `NewSlot` and `OnAttestation` are exported for fork choice spec tests

File: beacon-chain/blockchain/head.go [hunk 4]
```diff
@@ -37,21 +55,6 @@ func (s *Service) updateHead(ctx context.Context, balances []uint64) error {
 	ctx, span := trace.StartSpan(ctx, "blockChain.updateHead")
 	defer span.End()
 
-	// To get the proper head update, a node first checks its best justified
```
Comment (terencechain) [inline beacon-chain/blockchain/head.go:40]: This is an old implementation and should have been deleted a while ago

File: beacon-chain/blockchain/process_block_helpers.go [hunk 5]
```diff
@@ -328,56 +327,6 @@ func (s *Service) ancestorByDB(ctx context.Context, r [32]byte, slot types.Slot)
...
-//            # Update justified if new justified is later than store justified
-//            if state.current_justified_checkpoint.epoch > store.justified_checkpoint.epoch:
-//                store.justified_checkpoint = state.current_justified_checkpoint
-//                return
-//            # Update justified if store justified is not in chain with finalized checkpoint
-//            finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
-//            ancestor_at_finalized_slot = get_ancestor(store, store.justified_checkpoint.root, finalized_slot)
-//            if ancestor_at_finalized_slot != store.finalized_checkpoint.root:
-//                store.justified_checkpoint = state.current_justified_checkpoint
-func (s *Service) finalizedImpliesNewJustified(ctx context.Context, state state.BeaconState) error {
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block_helpers.go:346]: This is an old implementation that is deprecated via `NewSlot`

File: beacon-chain/blockchain/head.go [hunk 6]
```diff
@@ -23,6 +23,21 @@ import (
 	"go.opencensus.io/trace"
 )
 
+// UpdateHeadWithBalances updates the beacon state head after getting justified balanced from cache.
+func (s *Service) UpdateHeadWithBalances(ctx context.Context) error {
+	cp := s.store.JustifiedCheckpt()
+	if cp == nil {
```
Comment (rauljordan) [inline beacon-chain/blockchain/head.go:29]: See above ^. If we have a nil justified cpt, we return genesis, but here we return an error, which seems contradictory

File: beacon-chain/blockchain/process_block_test.go [hunk 7]
```diff
@@ -737,89 +736,6 @@ func TestEnsureRootNotZeroHashes(t *testing.T) {
 	assert.Equal(t, root, r, "Did not get wanted justified root")
 }
 
-func TestFinalizedImpliesNewJustified(t *testing.T) {
```
Comment (rauljordan) [inline beacon-chain/blockchain/process_block_test.go:740]: Is this logic still something that needs to be tested post-refactor? If so, where is it covered?

Comment (terencechain) [reply beacon-chain/blockchain/process_block_test.go:740]: nope, `finalizedImpliesNewJustified` was fully deprecated and removed

File: testing/spectest/shared/common/forkchoice/type.go [hunk 8]
```diff
@@ -0,0 +1,29 @@
+package forkchoice
```
[the remark this answers is not in the document]
Comment (terencechain) [reply testing/spectest/shared/common/forkchoice/type.go:1]: You are right, I followed the same pattern as `ssz_static` spec tests. We can address this all together in the following prs

**Review by rauljordan — DISMISSED** [review]

**Review by rauljordan — APPROVED** [review]

**Pull request merged**